### ISSUE TITLE
Tf.Variable handling when tf.function is set appears unfixably broken

### ISSUE THREAD

**Issue by jwlm-home** [tensorflow/tensorflow#43002]
-   **Have I written custom code (as opposed to using a stock example script
    provided in TensorFlow)**: Yes
-   **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: MacOS 10.15.6
-   **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue
    happens on a mobile device**: No
-   **TensorFlow installed from (source or binary)**: binary
-   **TensorFlow version (use command below)**: v2.3.0-rc2-23-gb36436b087 2.3.0
-   **Python version**: Python 3.7.4
-   **Bazel version (if compiling from source)**: N/A
-   **GCC/Compiler version (if compiling from source)**: N/A
-   **CUDA/cuDNN version**: N/A
-   **GPU model and memory**: N/A
-   **Exact command to reproduce**: run below script

### Describe the problem
Consider the following case:

(1) Running keras-style functional models
(2) Training requires that each evaluation step runs against several modified versions of the input data in which different columns are replaced at different times
(3) Running in eager mode is two orders of magnitude slower than running in graph mode, so running in eager mode is flat out.

### Source code / logs

(Approximate minimal case extracted from my source; the actual notebook is several thousand lines long and would merely confuse matters.)

import numpy as np
import tensorflow as tf

class BrokenLayer(tf.keras.layers.Layer):
    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        
    def call(self, inputs):
        transposed_inputs = tf.transpose(inputs)
        
        retval = []
        for delta in [[0], [0, 1]]:
            v = tf.Variable(
                np.zeros(len(delta, 1)),
                shape=tf.TensorShape((len(delta), None)))
            
            retval += [
                tf.tensor_scatter_nd_update(
                    tf.transpost(inputs),
                    delta,
                    v)]

        return retval

... No run this inside tf.keras.model.fit

It blows up because of the classic 'ValueError: tf.function-decorated function tried to create variables on non-first call.'

This behavior is by design, according to https://github.com/tensorflow/tensorflow/issues/27120. Unfortunately, despite spending the best part of a day trying to create a by-row constant tensor with a undefined second axis, I have found no way to replace the call to tf.Variable.

Lest it seem like this is a minor issue -- 'just create a Variable at class initialization and reuse that' is the first, obvous, suggestion -- *THAT DOES NOT WORK*. You need to be able to use tensor_scatter_nd_update to get there, and that's not possible, multiplying inputs by zero gives a tensor with the wrong shape. Using tensor array doesn't work, since it requires the each assignment be of known shape, and that brings us back to creating variables.

This design is completely broken, in what is not all that surprising a situation.

**Comment by bhack** [comment]
Is this still in the cluster of https://github.com/tensorflow/tensorflow/issues/42183#issuecomment-671522699?

**Comment by jwlm-home** [comment]
None of the work-arounds suggested in https://github.com/tensorflow/tensorflow/issues/27120 work.

**Comment by ravikyram** [comment]
@jwlm-home 

Can you please share colab link or simple standalone code with  proper indentation  and supporting files to reproduce the issue in our environment.It helps us in localizing the issue faster. Thanks!

**Comment by jwlm-home** [comment]
Colab link to working example:

https://colab.research.google.com/github/jwlm-home/variable_failure/blob/master/example.ipynb

Notice a couple of things:

(1) The call function within the custom layer is decorated with a @tf.function decorator. THat's because it will otherwise run in eager mode, in which case the second call to tf.Variable does not raise an error
(2) Other than the custom layer, this uses completely vanilla training; there's no custom training loop here.

**Comment by jwlm-home** [comment]
Also:

(3) To make this run completely correctly in eager mode, you need to add an extra if statement so that the variable initialization handles cases where the shape of the input tensor is fully specified.

**Comment by bhack** [comment]
I think that you can try to workaround like: 
https://github.com/tensorflow/tensorflow/issues/26812#issuecomment-473736883

See also @alextp rationale at https://github.com/tensorflow/tensorflow/issues/26812#issuecomment-474595550

**Comment by jwlm-home** [comment]
Incorrect. That depends on the tensors having known dimensions.

**Comment by bhack** [comment]
Is your colab working without `@tf.function`?

**Comment by jwlm-home** [comment]
As written, no. I'll add a cell which does work.

**Comment by jwlm-home** [comment]
OK, I've updated the code so it runs in eager mode but not in graph mode.

**Comment by bhack** [comment]
Your last formulation will fail also in both modes on `v = tf.Variable(..` with the upcoming version `pip install tf_nighlty`: 
```
    TypeError: An op outside of the function building code is being passed
    a "Graph" tensor. It is possible to have Graph tensors
    leak out of the function building context by including a
    tf.init_scope in your function building code.
    For example, the following function will fail:
      @tf.function
      def has_init_scope():
        my_constant = tf.constant(1.)
        with tf.init_scope():
          added = my_constant * 2
    The graph tensor has name: broken_layer_8/zeros:0
```

**Comment by jwlm-home** [comment]
Quite so. That's because removing the lambda thunk at line 22, yielding

        v = tf.Variable(
            replacement_values,
            shape=tf.TensorShape((len(delta), None)))

causes this error:

ValueError: Tensor-typed variable initializers must either be wrapped
in an init_scope or callable (e.g., `tf.Variable(lambda :
tf.truncated_normal([10, 40]))`) when building functions. Please file
a feature request if this restriction inconveniences you.


In other words, fixing the leaking graph error causes another error --
and it's exactly the one called out in the nightly build bug, up to
the 'init_scope' call.


Let me see what I can do to fix this bug.

**Comment by jwlm-home** [comment]
OK, fixed. It now gives the expected tf.variable creation error.

**Comment by bhack** [comment]
I think that in your formulation the problem is that you are trying to create the variable twice cause it is connected to the two branch of the conditional case for `replacement_values`. This will conflict the rules that you need to protect `tf.variable` against multiple evaluations:
> tf.function may evaluate your python function more than once.
What the RFC states instead is that you are allowed to create variables as long as variable creation only happens the first time your python function is evaluated.

**Comment by jwlm-home** [comment]
Yes, and that's exactly the bug. I need to be able to deal with the case
where I must update a set of columns in an input during training. That
requires using scatter_nd -- and that, in turn, requires a tf.variable to
handle the case where the input has unspecified length -- which is, of
course, the core case for fitting a model to a set of data in graph mode.
That, in turn, requires the creation of more than one tf.Variable during
training.

That is both forbidden (by design) and required (as explained above).
That's exactly the bug.

**Comment by ravikyram** [comment]
I have tried in colab with TF version 2.3, nightly version(`2.4.0-dev20200907`) and was able to reproduce the issue.Please, find the gist [here](https://colab.research.google.com/gist/ravikyram/e2e8f729e35d45c73ceb1c79c873ec20/untitled321.ipynb). Thanks!

**Comment by mdanatg** [comment]
`tf.scatter_nd_update` does work with Tensors. In general, there is no mathematical expression that should require a `tf.Variable` to implement.

However, there is a problem with the original code: you may not mix TF and NumPy ops. In other words, you need to use `tf.zeros` instead of `np.zeros`, `tf.constant` instead of `np.array`, `tf.reshape` instead of `x.reshape`, etc.

I'm including the corrected code below, which runs without error:

```
        delta = tf.constant([0])
        
        replacement_values = tf.zeros((len(delta), len(inputs)), dtype=tf.float32)
                    
        retval = tf.transpose(
            tf.tensor_scatter_nd_update(
                transposed_inputs,
                tf.reshape(delta, (-1, 1)),
                replacement_values))
```

**Comment by jwlm-home** [comment]
It's self-evident that you did not even bother to read the error. The error has nothing to do with whether or not tensors can be used -- in fact, the issue arises because tf. Variable is required to make the code work in the general case, when len(inputs) is unavailable due to being called with a tensor of indeterminate dimension, and tf.Variable can only be called once.

**Comment by mdanatg** [comment]
Again, tf.Variable is **not** required to handle indeterminate dimensions. So the premise of this issue is incorrect. If you have specific use-cases that you believe require a tf.Variable, feel free to include them here, although such questions are better asked on StackOverflow.
